### DISCUSSION TITLE
Portable executable - PyInstaller on macOS

### DISCUSSION THREAD

**Discussion by patoch01220** [pyinstaller/pyinstaller#5858] [Help]
Hello everyone, 

I have a small issue : I am creating a little django app for my sister and i'm developping it on macOS. I almost finished developping it, and i would like to create an executable so my sister doesn't have to run the commands by herself to start the localhost server etc. 

To do so I wanted to use pyInstaller and that's the first time i'm using it. 
So i've created a python script that start the localhost server and open a browser on the right url. 

This works well for me and everything is fine on my computer. 

The issue i have and i haven't thought about it at first is that when i tried using it on my sister's mac, it told me that it needed Xcode. 
It's kind of annoying for her to download it only for this little app. That's why i was looking for a way to include dev tools from Xcode or anything like that, but didn't find anything interesting on internet. 

Do you have any idea how i could avoid this issue ? 

Thank you very much for the time you took reading this, 
Patrick

**Comment by rokm** [comment]
> The issue i have and i haven't thought about it at first is that when i tried using it on my sister's mac, it told me that it needed Xcode.

Do you know where the XCode requirement comes from? The frozen application should not require it per-se, so this sounds like there's something in your code that's depending on XCode (could also be one of its utility programs, like something in your code spawning a `git` process, for example).

**Comment by patoch01220** [comment]
> > The issue i have and i haven't thought about it at first is that when i tried using it on my sister's mac, it told me that it needed Xcode.
> 
> Do you know where the XCode requirement comes from? The frozen application should not require it per-se, so this sounds like there's something in your code that's depending on XCode (could also be one of its utility programs, like something in your code spawning a `git` process, for example).

Thanks for your reply @rokm !

I did a very basic Django app that uses HTML, CSS, Python, and some panda to read CSV. 
I actually used also github to save my project, but i have the files git in my repo. Is it the problem ? 

More over, my python script first open a browser on the right url and then start the server with the command `python3 manage.py runserver' (this is the command to start the localhost server for django).
I've seen that opening the we page works well and did opened on my sister's computer. But this is starting the server that causes the issue.

Do you have any idea ? There is my python code in the manage.py file :

```
#!/usr/bin/env python
"""Django's command-line utility for administrative tasks."""
import os
import sys
import django


def main():
    """Run administrative tasks."""
    os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'Project_MACB.settings')
    try:
        from django.core.management import execute_from_command_line
    except ImportError as exc:
        raise ImportError(
            "Couldn't import Django. Are you sure it's installed and "
            "available on your PYTHONPATH environment variable? Did you "
            "forget to activate a virtual environment?"
        ) from exc
    execute_from_command_line(sys.argv)


if __name__ == '__main__':
    main()
```

**Comment by rokm** [comment]
>  I actually used also github to save my project, but i have the files git in my repo. Is it the problem ?

Nope. What I meant is, if you're using `git` module (from `GitPython`) within your python code, that would call `git` command, which would result in message that XCode (or XCode CLI tools) need to be installed.

Can you post the error message about requiring XCode (and any messages preceding it)? Hopefully that will shed some light on the context in which this is occurring...

**Comment by patoch01220** [comment]
Hmm okay, i see, but no i do not call git command in my app. 

Actually i've tried importing django in my script where i start the server and open the browser and now i have much more package and files in my dist folder, so i'll try now like this.

For the error i think it was something like Xcode is not present in the folder app, like the app wasn't found. My sister went back to work, I'll try now that i have new things and i'll get back to you when i'll have some news and also the exact error code.

Anyway, thaks a lot for your time !

**Comment by patoch01220** [comment]
> > I actually used also github to save my project, but i have the files git in my repo. Is it the problem ?
> 
> Nope. What I meant is, if you're using `git` module (from `GitPython`) within your python code, that would call `git` command, which would result in message that XCode (or XCode CLI tools) need to be installed.
> 
> Can you post the error message about requiring XCode (and any messages preceding it)? Hopefully that will shed some light on the context in which this is occurring...

So I did import django in my script and now it works better. But now i have another issue :  i have to use python3 to use django. 

So when i try executing the executable on my sister's mac, i have the error (The command line says) : 

```
xcode-select: note: no developer tools were found at '/Applications/Xcode.app', requesting install. Choose an option in the dialog to download the command line developer tool
```

And then i have a pop up window that tells me :
``` 
Some command line tools are needed for the command 'python3'. Would you like to install them?
```

I also tried using the command ```python3 -m PyInstaller myScript.py``` (so pyInstaller uses python3), but I still have the same issue as before. 
Do you have any ideas how to fix this problem ?
I mean is it possible to include or force pyInstaller to use python3 so i don't have to install it on my sister's mac ?

**Comment by rokm** [comment]
Ah yes, `python3` is now also part of XCode CLI tools.

The problem you have that something in your code (probably Django) is trying to spawn python3 interpreter. Which beats the point of freezing the application in the first place, as that python3 process will likely require full environment. So you'll need to find what exactly is going on and see if you can work around it.

(It could be the  `django.core.management.execute_from_command_line` that you seem to be using as dispatch in your main script).

**Comment by patoch01220** [comment]
> Ah yes, `python3` is now also part of XCode CLI tools.
> 
> The problem you have that something in your code (probably Django) is trying to spawn python3 interpreter. Which beats the point of freezing the application in the first place, as that python3 process will likely require full environment. So you'll need to find what exactly is going on and see if you can work around it.
> 
> (It could be the `django.core.management.execute_from_command_line` that you seem to be using as dispatch in your main script).

Yeah I actually use the command : ```python3 manage.py runserver``` to launch the server and this is this line that causes the problem i think. I mean it's due to this "python3" I think.

I've tried using only python but django needs to use python3 to be able to launch the server. 

Could i maybe activate the environement manually or maybe just simply include python3 in my PyInstaller or something like this ?
 
My script is pretty simple : 
```
import os
import sys 
import webbrowser
import django

webbrowser.open('http://127.0.0.1:8000/MACB') 
os.system('python3 /Volumes/PATOCH/Project_MACB/manage.py runserver')
```

**Comment by rokm** [comment]
PyInstaller does *not* bundle python interpreter executable. It bundles python's shared library, and uses that to set up embedded interpreter to run the frozen script (with its collected resources). If you want to spawn the interpreter, you will need to have it installed on your target machine. Along with the full environment you need to actually run the script you're trying to run. So in this case, you'd need to install python3 and django on your sister's Mac, and there's no point in using PyInstaller at all.

I think typically, you'd use PyInstaller to freeze your Django project and generate executable out of the `manage.py` script, and run the generated executable to start the site, e.g. `./dist/name/manage runserver --noreload`...

**Comment by patoch01220** [comment]
> PyInstaller does _not_ bundle python interpreter executable. It bundles python's shared library, and uses that to set up embedded interpreter to run the frozen script (with its collected resources). If you want to spawn the interpreter, you will need to have it installed on your target machine. Along with the full environment you need to actually run the script you're trying to run. So in this case, you'd need to install python3 and django on your sister's Mac, and there's no point in using PyInstaller at all.
> 
> I think typically, you'd use PyInstaller to freeze your Django project and generate executable out of the `manage.py` script, and run the generated executable to start the site, e.g. `./dist/name/manage runserver --noreload`...

Ohh yeah okay i see!

I just didn't really understand the last part (sorry i'm not fluent in english). Are you meaning that i could just use the same script as before but instead of calling ```python3 manage.py runserver```i could just replace that with the command you've entered (```./dist/name/manage runserver --noreload```)

Anyway thanks a lot for the time you're according to my issue, I appreciate it!

**Comment by rokm** [comment]
You definitely should be freezing your Django project, if you plan to deploy it elsewhere. Then you could probably run your simple script in old python2 as well, since all it'd be doing would be opening web browser via `webbrowser` and running the frozen executable via `os.system()`. 

What you were doing now, you were freezing your "launcher" script and some of Django modules, but none of your actual Django project... (On a side note, I'm not overly familiar with Django, much less with freezing it under PyInstaller, so I can't help you much with that).

**Comment by patoch01220** [comment]
Okay i see! 

Last question, do you know how to freeze my Django project ? 
What i mean is that you said : " I think typically, you'd use PyInstaller to freeze your Django project and generate executable out of the manage.py script, and run the generated executable to start the site, e.g. ./dist/name/manage runserver --noreload..."
But do you know how i could do that ? 

Because you've seen that when I want to start the server I have to call the file manage.py with the command runsever. 

Do you mean that i should apply the pyInstaller on the manyge.py file instead of my launcher script ?

**Comment by rokm** [comment]
> Do you mean that i should apply the pyInstaller on the manyge.py file instead of my launcher script ?

Yeah, I would try freezing the `manage.py` script.

**Comment by patoch01220** [comment]
Okay, i'll try checking all this out, 
Thanks a lot !
